src/Cli/dotnet/commands/dotnet-workload/install/FileBasedInstaller.cs

```diff
@@ -284,87 +295,162 @@ public IEnumerable<WorkloadDownload> GetDownloads(IEnumerable<WorkloadId> worklo
+            string workloadSetsDirectory = Path.Combine(manifestInstallDirForFeatureBand, SdkDirectoryWorkloadManifestProvider.WorkloadSetsFolderName);
+
+            foreach ((string workloadSetVersion, _) in installedWorkloadSets)
             {
-                return;
+                if (garbageCollector.WorkloadSetsToKeep.Contains(workloadSetVersion))
+                {
+                    //  Don't uninstall this workload set
+                    continue;
+                }
+
+                string workloadSetDirectory = Path.Combine(workloadSetsDirectory, workloadSetVersion);
+                if (Directory.Exists(workloadSetDirectory))
+                {
+                    //  If the directory doesn't exist, the workload set is probably from a directory specified via the DOTNETSDK_WORKLOAD_MANIFEST_ROOTS environment variable
+                    //  In that case just ignore it, as the CLI doesn't manage that install
+                    Directory.Delete(workloadSetDirectory, true);
+                }
             }
 
-            foreach (var packIdDir in Directory.GetDirectories(installedPacksDir))
+            //  Garbage collect workload manifests
+            Dictionary<(ManifestId manifestId, ManifestVersion manifestVersion, SdkFeatureBand manifestFeatureBand), List<SdkFeatureBand>> manifestInstallRecords = GetAllManifestInstallRecords();
+            foreach (var (manifestId, manifestVersion, manifestFeatureBand) in manifestInstallRecords.Keys)
             {
-                foreach (var packVersionDir in Directory.GetDirectories(packIdDir))
+                var referencingFeatureBands = manifestInstallRecords[(manifestId, manifestVersion, manifestFeatureBand)];
+                List<SdkFeatureBand> featureBandsToRemove = new();
+                foreach (var referencingFeatureBand in referencingFeatureBands)
                 {
-                    var bandPackRecordPaths = Directory.GetFileSystemEntries(packVersionDir);
+                    if (!installedSdkFeatureBands.Contains(referencingFeatureBand))
+                    {
+                        //  If an SDK feature band is no longer installed, manifests it references can be garbage collected
+                        featureBandsToRemove.Add(referencingFeatureBand);
+                    }
 
-                    var unneededBandRecordPaths = bandPackRecordPaths
-                        .Where(recordPath =>
-                            {
-                                var thisRecordFeatureBand = new SdkFeatureBand(Path.GetFileName(recordPath));
+                    if (referencingFeatureBand.Equals(_sdkFeatureBand) && !garbageCollector.ManifestsToKeep.Contains((manifestId, manifestVersion, manifestFeatureBand)))
+                    {
+                        //  For current feature band, garbage collect manifests that the garbage collector didn't mark as ones to keep
+                        featureBandsToRemove.Add(referencingFeatureBand);
+                    }
```

Comment:
Consider combining these into 1 if-statement with an `||` between the predicates.

Reply:
I left them separate because it seems like the comments make it clearer to understand that way.